### PR TITLE
Laboratory work #4, Christina Nigamedzyanova - 21FPL2

### PR DESCRIPTION
**Pull request by christine-ni**

### CODE REVIEW

File: lab_4_summarization_textrank/main.py [hunk 1]
```diff
@@ -120,23 +147,33 @@ def __init__(self) -> None:
...
+                self._id2word[count] = token
 
     def encode_sentences(self, sentences: tuple[Sentence, ...]) -> None:
         """
         Enriches the instances of sentences with their encoded versions
         :param sentences: a sequence of sentences
         :return: a list of sentences with their preprocessed versions
         """
-        pass
+        #я запуталась здесь
```
Comment (marina-kaz) [inline lab_4_summarization_textrank/main.py:170]: You want to firstly check the validity of the arguments, then access its preprocessed version, then compose its encoded version using the corresponding attribute, and this should be it

**Review by marina-kaz — COMMENTED** [review]
That's a great start!
Keep going

File: lab_4_summarization_textrank/main.py [hunk 2]
```diff
@@ -120,23 +147,33 @@ def __init__(self) -> None:
...
         """
         Fills attributes mapping words and integer equivalents to each other
         :param tokens: a sequence of string tokens
         :return:
         """
-        pass
+        for count, token in enumerate(tokens, start=1000):
+            if token not in (self._word2id.keys() and self._id2word.keys()):
+                self._word2id[token] = count
+                self._id2word[count] = token
```
[the remark this answers is not in the document]
Comment (christine-ni) [reply lab_4_summarization_textrank/main.py:162]: done

File: lab_4_summarization_textrank/main.py (hunk 1 above)
Comment (christine-ni) [reply lab_4_summarization_textrank/main.py:170]: done

File: lab_4_summarization_textrank/main.py [hunk 3]
```diff
@@ -84,32 +93,50 @@ def __init__(self, stop_words: tuple[str, ...], punctuation: tuple[str, ...]) ->
...
+
 
     def _preprocess_sentences(self, sentences: tuple[Sentence, ...]) -> None:
         """
         Enriches the instances of sentences with their preprocessed versions
         :param sentences: a list of sentences
         :return:
         """
-        pass
+        for sentence in sentences:
```
[the remark this answers is not in the document]
Comment (christine-ni) [reply lab_4_summarization_textrank/main.py:126]: done

File: lab_4_summarization_textrank/main.py [hunk 4]
```diff
@@ -146,8 +183,15 @@ def calculate_similarity(sequence: Union[list, tuple], other_sequence: Union[lis
     :param other_sequence: a sequence of items
     :return: similarity score
     """
-    pass
-
+    if not sequence or not other_sequence:
```
[the remark this answers is not in the document]
Comment (christine-ni) [reply lab_4_summarization_textrank/main.py:186]: done

File: lab_4_summarization_textrank/main.py [hunk 5]
```diff
@@ -160,14 +204,15 @@ def __init__(self) -> None:
         """
         Constructs necessary attributes
         """
-        pass
+        self._matrix = []
+        self.vertices = []
```
[the remark this answers is not in the document]
Comment (christine-ni) [reply lab_4_summarization_textrank/main.py:208]: i forgot about protecting it, now i turned it into a protected argument

File: lab_3_keywords_textrank/main.py [hunk 6]
```diff
@@ -205,14 +206,24 @@ def extract_pairs(tokens: tuple[int, ...], window_length: int) -> Optional[tuple
...
+    for count, token in enumerate(tokens):
+        if count < len(tokens) - window_length + 1:
+            for i in range(1, window_length):
+                if token != tokens[count + i]:
+                    pairs.append(tuple([token, tokens[count + i]]))
+        else:
+            for i in range(len(tokens) - count):
+                if token != tokens[count + i]:
+                    pairs.append(tuple([token, tokens[count + i]]))
+    return tuple(pairs)
```
[the remark this answers is not in the document]
Comment (christine-ni) [reply lab_3_keywords_textrank/main.py:219]: as far as i remember, i changed this function a bit when i had a defense with Artyom Mikhailovich as he advised me to do that

File: lab_3_keywords_textrank/main.py [hunk 7]
```diff
@@ -671,10 +683,11 @@ def get_top_keywords(self, n_keywords: int) -> tuple[int, ...]:
             tuple[int, ...]
                 top n most important tokens in the encoded text
         """
+        return tuple(k for k, v in sorted(self._scores.items(), key=lambda item: (item[1], -item[0]),
```
[the remark this answers is not in the document]
Comment (christine-ni) [reply lab_3_keywords_textrank/main.py:686]: i don't remember changin this string....

File: lab_3_keywords_textrank/main.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (marina-kaz) [reply lab_3_keywords_textrank/main.py:219]: after lab defense your fork has been updated
as of today, in this folder it is not your implementation but the one of your groupmate
so please discard any changes in this file

File: lab_4_summarization_textrank/main.py [hunk 8]
```diff
@@ -84,59 +113,99 @@ def __init__(self, stop_words: tuple[str, ...], punctuation: tuple[str, ...]) ->
...
+        sentences = self._split_by_sentence(text)
+        self._preprocess_sentences(sentences)
+        return sentences
 
 class SentenceEncoder(TextEncoder):
     """
     A class to encode string sequence into matching integer sequence
     """
 
-    def __init__(self) -> None:
```
Comment (marina-kaz) [inline lab_4_summarization_textrank/main.py:119]: consider using init for keeping track of the max identificator

File: lab_4_summarization_textrank/main.py [hunk 9]
```diff
@@ -84,59 +113,99 @@ def __init__(self, stop_words: tuple[str, ...], punctuation: tuple[str, ...]) ->
...
         """
-        pass
+        if not check_type(sentences, tuple, Sentence):
+            raise ValueError
+        words = []
+        for sentence in sentences:
+            prepr = sentence.get_preprocessed()
+            for word in prepr:
+                words.append(word)
+        self._learn_indices(tuple(words))
```
Comment (marina-kaz) [inline lab_4_summarization_textrank/main.py:202]: given the fact you can call `_learn_indices` more than once, consider calling it from each of the sentences

File: lab_4_summarization_textrank/main.py [hunk 10]
```diff
@@ -146,8 +215,20 @@ def calculate_similarity(sequence: Union[list, tuple], other_sequence: Union[lis
...
+    if not isinstance(other_sequence, (list, tuple)):
+        raise ValueError
+    if not sequence or not other_sequence:
+        return 0
+    numerator = []
+    denominator = set(tuple(sequence) + tuple(other_sequence))
+    for i in sequence:
+        if i in other_sequence:
+            numerator.append(i)
+    set(numerator)
```
Comment (marina-kaz) [inline lab_4_summarization_textrank/main.py:229]: `set` does not remove duplicates inplace

**Review by marina-kaz — COMMENTED** [review]
Very good!
A bit of room for improvement

File: lab_4_summarization_textrank/main.py (hunk 10 above)
Comment (christine-ni) [reply lab_4_summarization_textrank/main.py:229]: I'm sorry, I don't understand what do you want me to do...
should i delete duplicates in sequence and other_sequence before adding elements to numerator?

File: lab_4_summarization_textrank/main.py (hunk 9 above)
Comment (christine-ni) [reply lab_4_summarization_textrank/main.py:202]: done (?)

File: lab_4_summarization_textrank/main.py [hunk 11]
```diff
@@ -84,59 +111,99 @@ def __init__(self, stop_words: tuple[str, ...], punctuation: tuple[str, ...]) ->
...
         Fills attributes mapping words and integer equivalents to each other
         :param tokens: a sequence of string tokens
         :return:
         """
-        pass
+        if not check_type(tokens, tuple, str):
+            raise ValueError
+        for count, token in enumerate(tokens, start=self._max_identificator):
+            if token not in (self._word2id.keys() and self._id2word.keys()):
+                self._word2id[token] = count
```
Comment (marina-kaz) [inline lab_4_summarization_textrank/main.py:183]: why check if token in id2word?

File: lab_4_summarization_textrank/main.py (hunk 9 above)
Comment (marina-kaz) [reply lab_4_summarization_textrank/main.py]: not exaclty
I meant that you can just `_learn_indices(sentence.get_preprocessed())`

File: lab_4_summarization_textrank/main.py (hunk 10 above)
Comment (marina-kaz) [reply lab_4_summarization_textrank/main.py]: it is ok now, thanks!

**Review by marina-kaz — COMMENTED** [review]
Good job!
Consider aiming for a higher mark?

**Comment by christine-ni** [comment]
> Good job! Consider aiming for a higher mark?

I would like to, but unfirtunately i don't have time because of the deadlines and upcoming exams(

File: lab_4_summarization_textrank/main.py (hunk 9 above)
Comment (christine-ni) [reply lab_4_summarization_textrank/main.py]: done

**Pull request closed without merging**